### PR TITLE
KAFKA-12829: Remove deprecated Topology#addProcessor of old Processor API

### PR DESCRIPTION
**Pull request by mjsax**
Must be cherry-picked to `4.0` branch.

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java [hunk 1]
```diff
@@ -57,7 +57,7 @@
 import org.apache.kafka.test.GenericInMemoryKeyValueStore;
 import org.apache.kafka.test.MockApiProcessor;
 import org.apache.kafka.test.MockApiProcessorSupplier;
-import org.apache.kafka.test.MockInternalNewProcessorContext;
+import org.apache.kafka.test.MockInternalProcessorContext;
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:60]: Follow up clean from https://github.com/apache/kafka/pull/18103

Rename this class, what result in lot of updates in many files...

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamProcessorApiTest.java [hunk 2]
```diff
@@ -46,7 +46,7 @@
 import static java.util.Arrays.asList;
 
 
-public class KStreamNewProcessorApiTest {
+public class KStreamProcessorApiTest {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamProcessorApiTest.java:49]: This is a similar update -- we don't have the old Processor API any longer, so dropping the `New` from this name, too.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java [hunk 3]
```diff
@@ -109,8 +110,8 @@ public void setup() {
         // Create a new directory in which we'll put all of the state for this test, enabling running tests in parallel ...
         final File localState = TestUtils.tempDirectory();
         props.setProperty(StreamsConfig.STATE_DIR_CONFIG, localState.getAbsolutePath());
-        props.setProperty(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.String().getClass().getName());
-        props.setProperty(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.String().getClass().getName());
+        props.setProperty(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.StringSerde.class.getName());
+        props.setProperty(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.StringSerde.class.getName());
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:114]: Side cleanup to get rid of some IntelliJ warning. (also cleaner code.

Similar stuff below.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java [hunk 4]
```diff
@@ -337,39 +338,7 @@ public void testDrivingConnectedStateStoreTopology() {
...
-        inputTopic.pipeInput("key3", "value3");
-        inputTopic.pipeInput("key1", "value4");
-        assertTrue(outputTopic1.isEmpty());
-
-        final KeyValueStore<String, String> store = driver.getKeyValueStore("connectedStore");
-        assertEquals("value4", store.get("key1"));
-        assertEquals("value2", store.get("key2"));
-        assertEquals("value3", store.get("key3"));
-        assertNull(store.get("key4"));
-    }
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:353]: Diff is just weird.. this was not removed

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java [hunk 5]
```diff
@@ -337,39 +338,7 @@ public void testDrivingConnectedStateStoreTopology() {
...
-
-    @Deprecated // testing old PAPI
-    @Test
-    public void testDrivingConnectedStateStoreInDifferentProcessorsTopologyWithOldAPI() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:357]: removed this (and some other tests) -- verified that we have equivalent tests for new Processor API for all of them

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java [hunk 6]
```diff
@@ -815,439 +784,7 @@ public void testPrefixScanLruMapWithCachingWithLogging() {
...
-        final List<KeyValue<String, String>> results = prefixScanResults(store, DEFAULT_PREFIX);
-
-        assertEquals("key1", results.get(0).key);
-        assertEquals("value4", results.get(0).value);
-        assertEquals("key2", results.get(1).key);
-        assertEquals("value2", results.get(1).value);
-        assertEquals("key3", results.get(2).key);
-        assertEquals("value3", results.get(2).value);
-
-    }
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:836]: Weird diff..

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java [hunk 7]
```diff
@@ -1568,21 +1105,19 @@ private Topology createSimpleTopologyWithDroppingPartitioner() {
                 .addSink("sink", OUTPUT_TOPIC_1, new DroppingPartitioner(), "processor");
     }
 
-    @Deprecated // testing old PAPI
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:1571]: This annotation was wrong -- should have been `@SuppressWarning("deprecation")`

Updating this code to use new Processor API now.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java [hunk 8]
```diff
@@ -1568,21 +1105,19 @@ private Topology createSimpleTopologyWithDroppingPartitioner() {
...
     private Topology createStatefulTopology(final String storeName) {
         return topology
             .addSource("source", STRING_DESERIALIZER, STRING_DESERIALIZER, INPUT_TOPIC_1)
-            .addProcessor("processor", define(new OldAPIStatefulProcessor(storeName)), "source")
+            .addProcessor("processor", () -> new StatefulProcessor(storeName), "source")
             .addStateStore(Stores.keyValueStoreBuilder(Stores.inMemoryKeyValueStore(storeName), Serdes.String(), Serdes.String()), "processor")
             .addSink("counts", OUTPUT_TOPIC_1, "processor");
     }
 
-    @Deprecated // testing old PAPI
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:1580]: as above

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java [hunk 9]
```diff
@@ -1729,30 +1264,6 @@ public void process(final Record<String, String> record) {
         }
     }
 
-    /**
-     * A processor that stores each key-value pair in an in-memory key-value store registered with the context.
-     */
-    @SuppressWarnings("deprecation") // Old PAPI. Needs to be migrated.
-    protected static class OldAPIStatefulProcessor extends org.apache.kafka.streams.processor.AbstractProcessor<String, String> {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:1736]: There is already a `StatefulProcessor` for the new PAPI -- nothing to be migrated -- deleting.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java [hunk 10]
```diff
@@ -1775,30 +1286,9 @@ public void process(final Record<String, String> record) {
         }
     }
 
-    @SuppressWarnings("deprecation") // Old PAPI. Needs to be migrated.
-    private <K, V> org.apache.kafka.streams.processor.ProcessorSupplier<K, V> define(final org.apache.kafka.streams.processor.Processor<K, V> processor) {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:1779]: just dropped this one -- not really useful to have..

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java [hunk 11]
```diff
@@ -1775,30 +1286,9 @@ public void process(final Record<String, String> record) {
...
-        return () -> processor;
-    }
-
-    @SuppressWarnings("deprecation") // Old PAPI. Needs to be migrated.
-    private <K, V> org.apache.kafka.streams.processor.ProcessorSupplier<K, V> defineWithStoresOldAPI(final Supplier<org.apache.kafka.streams.processor.Processor<K, V>> supplier,
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:1784]: There is already `defineWithStores()` for new PAPI -- nothing to be migrated -- deleting

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java [hunk 12]
```diff
@@ -1775,30 +1286,9 @@ public void process(final Record<String, String> record) {
...
-            public Set<StoreBuilder<?>> stores() {
-                return stores;
-            }
-        };
-    }
-
     private <KIn, VIn, KOut, VOut> ProcessorSupplier<KIn, VIn, KOut, VOut> defineWithStores(final Supplier<Processor<KIn, VIn, KOut, VOut>> supplier,
                                                                                             final Set<StoreBuilder<?>> stores) {
-        return new ProcessorSupplier<KIn, VIn, KOut, VOut>() {
+        return new ProcessorSupplier<>() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:1291]: side cleanup

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java [hunk 13]
```diff
@@ -1568,21 +1105,19 @@ private Topology createSimpleTopologyWithDroppingPartitioner() {
                 .addSink("sink", OUTPUT_TOPIC_1, new DroppingPartitioner(), "processor");
     }
 
-    @Deprecated // testing old PAPI
     private Topology createStatefulTopology(final String storeName) {
         return topology
             .addSource("source", STRING_DESERIALIZER, STRING_DESERIALIZER, INPUT_TOPIC_1)
-            .addProcessor("processor", define(new OldAPIStatefulProcessor(storeName)), "source")
+            .addProcessor("processor", () -> new StatefulProcessor(storeName), "source")
```
Comment (pegasas) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTest.java:1111]: Indeed, it looks like `OldAPIStatefulProcessor` has been corrected to `StatefulProcessor` for migrating to current API. Got it! Thanks for teaching!

**Review by pegasas — APPROVED** [review]

**Comment by chia7712** [comment]
@mjsax could you please fix the conflicts ?

File: streams/test-utils/src/test/java/org/apache/kafka/streams/MockProcessorContextTest.java [hunk 14]
```diff
@@ -1,395 +0,0 @@
...
-import static org.junit.jupiter.api.Assertions.assertTrue;
-import static org.junit.jupiter.api.Assertions.fail;
-import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.ArgumentMatchers.anyString;
-import static org.mockito.Mockito.doAnswer;
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.when;
-
-@SuppressWarnings("deprecation") // this is a test of a deprecated API
-public class MockProcessorContextTest {
```
Comment (chia7712) [inline streams/test-utils/src/test/java/org/apache/kafka/streams/MockProcessorContextTest.java:57]: The deprecated `org.apache.kafka.streams.processor.MockProcessorContext` still exists in the codebase. Should we retain the unit tests?

**Review by chia7712 — APPROVED** [review]
@mjsax thanks for this patch. LGTM and one small question is left.

**Review by bbejeck — APPROVED** [review]
Thanks for the PR @mjsax - LGTM

File: streams/test-utils/src/test/java/org/apache/kafka/streams/MockProcessorContextTest.java (hunk 14 above)
Comment (lucasbru) [reply streams/test-utils/src/test/java/org/apache/kafka/streams/MockProcessorContextTest.java:57]: Good question. I suppose the trouble is that the unit test uses the removed classes, but the unit under test will not be removed yet. Can the tests be rewritten with transformers?

**Review by lucasbru — APPROVED** [review]
Yes, LGTM modulo the question raised by Chia

File: streams/test-utils/src/test/java/org/apache/kafka/streams/MockProcessorContextTest.java (hunk 14 above)
Comment (mjsax) [reply streams/test-utils/src/test/java/org/apache/kafka/streams/MockProcessorContextTest.java:57]: Good catch.

**Comment by mjsax** [comment]
Java 23:
```
Found 1 test failures:
FAILED ❌ PlaintextAdminIntegrationTest > testConsumerGroupsDeprecatedConsumerGroupState(String, String).quorum=kraft.groupProtocol=consumer
Found 1 flaky test failures:
FLAKY ⚠️  SslTransportLayerTest > "testSelectorPollReadSize(Args).args=tlsProtocol=TLSv1.2, useInlinePem=false"
```

Java 17:
```
Found 5 test failures:
FAILED ❌ AuthorizerIntegrationTest > testConsumeUsingAssignWithNoAccess(String, String).quorum=kraft.groupProtocol=consumer
FAILED ❌ PlaintextProducerSendTest > testCloseWithZeroTimeoutFromCallerThread(String, String).quorum=kraft.groupProtocol=classic
FAILED ❌ RemoteLogManagerTest > testRLMOpsWhenMetadataIsNotReady()
FAILED ❌ PlaintextAdminIntegrationTest > testConsumerGroupsDeprecatedConsumerGroupState(String, String).quorum=kraft.groupProtocol=consumer
FAILED ❌ PlaintextAdminIntegrationTest > testConsumerGroups(String, String).quorum=kraft.groupProtocol=consumer
Found 2 flaky test failures:
FLAKY ⚠️  TransactionsWithTieredStoreTest > "testBumpTransactionalEpochWithTV2Enabled(String, String, boolean).quorum=kraft, groupProtocol=consumer, isTV2Enabled=true"
FLAKY ⚠️  SslTransportLayerTest > "testSelectorPollReadSize(Args).args=tlsProtocol=TLSv1.2, useInlinePem=false"
```

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `4.0` branch.

**Pull request merged**